### PR TITLE
copycat: init at 001

### PR DESCRIPTION
**Pull request by DeeKahy**
## Description of changes

Add CopyCat `ccat` for walking a project tree to copy all text readable files to the clipboard with file names and file locations.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/misc/CopyCat/ccat.nix
Comment (eclairevoyant) [inline pkgs/tools/misc/CopyCat/ccat.nix:12]:
```suggestion
    hash = "sha256-zllxQifRMNEMa3RO5WKrwGAUf1xQg6YrQBzIHzy43F0=";
```

File: pkgs/tools/misc/CopyCat/ccat.nix
Comment (eclairevoyant) [inline pkgs/tools/misc/CopyCat/ccat.nix:16]:
```suggestion
  cargoHash = "sha256-LYVhvq5l+PCZXW+elWi3zZFxLekgPn+plo4dybbLK9g=";
```

File: pkgs/tools/misc/CopyCat/ccat.nix
Comment (eclairevoyant) [inline pkgs/tools/misc/CopyCat/ccat.nix:22]:
```suggestion
  meta = {
    description = "Utility to copy project tree contents to clipboard";
    homepage = "https://github.com/DeeKahy/CopyCat";
    license = lib.licenses.mit;
    maintainers = [ lib.maintainers.deekahy ];
    mainProgram = "ccat";
```

File: pkgs/tools/misc/CopyCat/ccat.nix
Comment (eclairevoyant) [inline pkgs/tools/misc/CopyCat/ccat.nix:10]:
```suggestion
    rev = "refs/tags/${version}";
```

File: pkgs/tools/misc/CopyCat/ccat.nix
Comment (eclairevoyant) [inline pkgs/tools/misc/CopyCat/ccat.nix:1]: Please move this to `pkgs/by-name/co/copycat/package.nix`
you may need to nixfmt as well

**Review by eclairevoyant — DISMISSED** [review]

**Comment by eclairevoyant** [comment]
Also, please squash and update the commit message to `copycat: init at 001` (case matters!)

File: pkgs/tools/misc/CopyCat/ccat.nix
Comment (eclairevoyant) [reply pkgs/tools/misc/CopyCat/ccat.nix:1]: No need to apologise. The PR reviews are here for a reason, your contribution is much appreciated.

**Review by eclairevoyant — COMMENTED** [review]
This builds and looks fine, but when I run it in a small-ish git repo I do seem to get a lot of broken pipe errors: https://gist.github.com/eclairevoyant/cc9771406a86c71b78eddbeb241350d5

Also it seems to be copying the files in `.git`, is that expected?

**Comment by DeeKahy** [comment]
> Also it seems to be copying the files in .git, is that expected?
If it prints `Found .gitignore file, adding its contents to the blacklist` and the gitignore contains `/.git` then it should not copy the contents of the .git folder. Otherwise yes it would add them to the clipboard.

> when I run it in a small-ish git repo I do seem to get a lot of broken pipe errors: https://gist.github.com/eclairevoyant/cc9771406a86c71b78eddbeb241350d5

This happens sometimes. I'm assuming it has something to do with a file currently being open in another application, but I haven't been able to find a fix for it yet.
Does it still happen in a different smallish project that isn't currently running or open in any editor?

**Comment by eclairevoyant** [comment]
That's fine, I just wanted to make sure it was an upstream issue and not an artifact of the packaging for nixpkgs.

One thing I missed: the PR needs one additional commit (coming first):
`maintainers: add deekahy`
with the contents following these guidelines: https://github.com/NixOS/nixpkgs/tree/master/maintainers#individual-maintainer-list

You can also look at other merged first-time-contribution PRs for examples.

**Review by eclairevoyant — APPROVED** [review]

**Comment by eclairevoyant** [comment]
will merge if CI doesn't come back with any flags

**Pull request merged**
